docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,387 @@
+updated schema, such as converting `STRING -> BYTES` or `STRING(10) ->
+STRING(25)`. These enable a fast-path conversion by simply updating the
+table descriptor.
+
+### Validated
+
+Validated conversions involve no conversion of on-disk data, but do
+require that all existing data is validated against the proposed schema
+change before completion. This would include conversions such as
+`STRING(100) -> STRING(10)` or `BYTES -> STRING`.
+
+In cases where existing data conflicts with the requested type, the type
+conversion will fail, rather than attempting to automatically coerce
+non-conforming data. In the example of `STRING(100) -> STRING(10)`
+above, the user would have the option of using an assisted cast
+(described below) to achieve the desired result e.g. `USING
+substring(my_string_column, 1, 10)`.
+
+These have a cost somewhere between trivial and general conversions,
+since on-disk data does not need to be rewritten, only revalidated.
+
+### Cast
+
+Cast conversions involve applying a built-in cast behavior to the
+existing column data requiring the on-disk KV data to be rewritten, such
+as `UUID -> STRING`.
+
+### Assisted
+
+Assisted conversions involve applying a `USING expression` to the
+existing column data, such as `FLOAT -> INT USING ceil(column_name)`.
+The allowable expressions are the same as those used in
+[computed columns](https://www.cockroachlabs.com/docs/stable/computed-columns.html#creation),
+namely pure functions using the data available in a single row.
+
+## Two-phase change
+
+An `ALTER ... COLUMN TYPE` command is implemented in two phases:
+1. Add a new, hidden, computed column to the table that effects the
+   conversion.
+2. Atomically remove the computed nature from the new column, swap it
+   for the original, and drop the original column.
+
+Given that backfilling the computed column may take an arbitrary and
+unpredictable amount of time, we will document and support a "manual"
+version of the above, allowing a user to control exactly when a type
+change becomes visible to SQL clients.
+
+## Key constraints
+
+Columns that are the target of a primary or foreign key constraint will
+not be supported under this RFC. It could be possible to support foreign
+keys in the future by cascading the type change, but this would greatly
+increase the amount of coordination required to effect the total schema
+change.
+
+### Interleaved tables
+
+Changing the types of primary keys is not within the scope of this RFC,
+so this feature does not interact with interleaved tables. When changes
+to primary keys are supported, a change to the PK of a parent table
+would have to be cascaded to the corresponding PK prefix column of any
+interleaved descendant tables.
+
+## Indexed columns
+
+Columns that are part of an index are supported, however columns that
+are part of a primary key are explicitly out of scope for this change.
+
+## Views
+
+Columns that are depended upon by a `CREATE VIEW` are out of scope for
```

Comment:
Clarify that the changes will ensure that a client receives an error if they attempt to chang ethe type of a column depended on by a view.

Reply:
Also indexes as a whole can be depended on by something else (not just FKs). 

Try this: "In general any time there is a structural dependency on a column or an index, the alter proposed here will bail out with an error. Supporting these cases is left to later work."